src/Microsoft.AspNetCore.Authentication/AuthenticationPolicyHandler.cs

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Authentication
+{
+    public class AuthenticationPolicyOptions : AuthenticationSchemeOptions {
+        // Used as the fallback if any of the explicit schemes and the default scheme selected is null.
+        public string DefaultScheme { get; set; }
+
+        public string AuthenticateScheme { get; set; }
+        public string ChallengeScheme { get; set; }
+        public string ForbidScheme { get; set; }
+        public string SignInScheme { get; set; }
+        public string SignOutScheme { get; set; }
+
+        /// <summary>
+        /// Used to 
+        /// </summary>
+        public Func<HttpContext, string> DefaultSchemeSelector { get; set; }
```

Comment:
Isn't this a policy selector?

Reply:
Well this one is actually selecting what scheme to resolve to for this policy instance, as opposed to a policy selector.  

But naming is getting confusing here, since "policies" are just special authentication options that contain a full set of schemes to redirect the normal `IAuthenticationHandler` calls to...

Reply:
I think it's better to give a us a general description on the concepts that are necessary and complete and come naturally. Like given a restricted resource, be it code, data, or anything else, it maps to "policy" name(string). Then in AuthorizationOptions, policy name maps to an AuthorizationPolicy that mainly list of requirements...

Reply:
Yeah so that's exactly why I don't think we should call this new thing Policy, as its totally not the same thing as Authorization policy today.  We need a new word to use to disambiguate...

Reply:
I'd suggest a weirder name like calling it `AddSchemeSet/Map()` instead of `AddPolicy`